cmd/snap-bootstrap/partition/encrypt.go

```diff
@@ -133,3 +161,61 @@ func cryptsetupClose(name string) error {
 	}
 	return nil
 }
+
+func cryptsetupAddKey(key EncryptionKey, rkey RecoveryKey, node string) error {
+	// create a named pipe to pass the recovery key
+	fpath := filepath.Join(dirs.SnapRunDir, "tmp-rkey")
+	if err := syscall.Mkfifo(fpath, 0600); err != nil {
+		return fmt.Errorf("cannot create named pipe: %v", err)
+	}
+	defer os.RemoveAll(fpath)
+
+	// add a new key to slot 1 reading the passphrase from the named pipe
+	args := []string{
+		// add a new key
+		"luksAddKey",
+		// the encrypted device
+		node,
+		// batch processing, no password verification
+		"-q",
+		// read existing key from stdin
+		"--key-file", "-",
+		// store it in keyslot 1
+		"--key-slot", "1",
+		// the named pipe
+		fpath,
+	}
+
+	cmd := exec.Command("cryptsetup", args...)
+	cmd.Stdin = bytes.NewReader(key[:])
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+
+	// open the named pipe and write the recovery key
+	file, err := os.OpenFile(fpath, os.O_WRONLY, 0600)
+	if err != nil {
+		return fmt.Errorf("cannot open recovery key pipe: %v", err)
+	}
+	n, err := file.Write(rkey[:])
+	if n != recoveryKeySize {
+		file.Close()
+		return fmt.Errorf("cannot write recovery key: short write (%d bytes written)", n)
+	}
+	if err != nil {
```

Comment:
Not knowing anything about `cryptsetup`, is it safe to kill mid-flight? Would it be better to process a sequence of  SIGTERM/SIGQUIT/SIGKILL to give the program a chance to do whatever it needs to, to quit gracefully?

Reply:
That handling is performed in the unlikely case where we can't write to the named pipe. In that situation we won't be able to add a recovery key and the installation is not recoverable. We're killing the process in such a drastic manner to ensure we won't be blocked at `cmd.Wait()`. Any leftovers resulting from the abrupt termination shouldn't be relevant, even encryption keys in temporary files, because the system will need to be reinstalled. But we can do it in a more graceful way if you strongly feel it's necessary.